File: website/vo2/kt/src/wasmJsMain/kotlin/Timestamp.kt
// Parses ISO 8601 like "2024-01-15T10:30:00Z" or "2024-01-15T10:30:00.000Z"
// into seconds since 2000-01-01. Returns null on any parse / range error.

private fun isLeap(year: Int): Boolean =
    (year % 4 == 0 && year % 100 != 0) || year % 400 == 0

private val MONTH_DAYS = intArrayOf(31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)

internal fun parseTimestamp(input: String): Double? {
    var s = input.trim()
    if (s.endsWith('Z')) s = s.dropLast(1)
    val t = s.indexOf('T')
    if (t < 0) return null
    val dateStr = s.substring(0, t)
    var timeStr = s.substring(t + 1)

    val dp = dateStr.split('-')
    if (dp.size != 3) return null
    val year = dp[0].toIntOrNull() ?: return null
    val month = dp[1].toIntOrNull() ?: return null
    val day = dp[2].toIntOrNull() ?: return null

    val dot = timeStr.indexOf('.')
    if (dot >= 0) timeStr = timeStr.substring(0, dot)
    val tp = timeStr.split(':')
    if (tp.size != 3) return null
    val hour = tp[0].toDoubleOrNull() ?: return null
    val min = tp[1].toDoubleOrNull() ?: return null
    val sec = tp[2].toDoubleOrNull() ?: return null

    if (year !in 2000..2100) return null
    if (month !in 1..12) return null
    val maxDay = when (month) {
        2 -> if (isLeap(year)) 29 else 28
        4, 6, 9, 11 -> 30
        else -> 31
    }
    if (day !in 1..maxDay) return null
    if (hour < 0 || hour > 23 || min < 0 || min > 59 || sec < 0 || sec > 59) return null

    var days = 0
    for (y in 2000 until year) days += if (isLeap(y)) 366 else 365
    for (m in 0 until (month - 1)) {
        days += MONTH_DAYS[m]
        if (m == 1 && isLeap(year)) days += 1
    }
    days += day - 1

    return days * 86400.0 + hour * 3600.0 + min * 60.0 + sec
}

File: website/vo2/js/vo2/timestamp.js
// ISO 8601 timestamp parser. Returns seconds since 2000-01-01, or null.
// Handles "2024-01-15T10:30:00Z" and "2024-01-15T10:30:00.000Z".

function isLeap(year) {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}

const MONTH_DAYS = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

export function parseTimestamp(s) {
  s = s.trim();
  if (s.endsWith('Z')) s = s.slice(0, -1);

  const t = s.indexOf('T');
  if (t < 0) return null;
  const dateStr = s.slice(0, t);
  let timeStr = s.slice(t + 1);

  const dp = dateStr.split('-');
  if (dp.length !== 3) return null;
  const year = parseInt(dp[0], 10);
  const month = parseInt(dp[1], 10);
  const day = parseInt(dp[2], 10);
  if (Number.isNaN(year) || Number.isNaN(month) || Number.isNaN(day)) return null;

  const dot = timeStr.indexOf('.');
  if (dot >= 0) timeStr = timeStr.slice(0, dot);
  const tp = timeStr.split(':');
  if (tp.length !== 3) return null;
  const hour = parseFloat(tp[0]);
  const min = parseFloat(tp[1]);
  const sec = parseFloat(tp[2]);
  if (Number.isNaN(hour) || Number.isNaN(min) || Number.isNaN(sec)) return null;

  if (year < 2000 || year > 2100) return null;
  if (month < 1 || month > 12) return null;
  let maxDay;
  if (month === 2) maxDay = isLeap(year) ? 29 : 28;
  else if (month === 4 || month === 6 || month === 9 || month === 11) maxDay = 30;
  else maxDay = 31;
  if (day < 1 || day > maxDay) return null;
  if (hour < 0 || hour > 23 || min < 0 || min > 59 || sec < 0 || sec > 59) return null;

  let days = 0;
  for (let y = 2000; y < year; y++) days += isLeap(y) ? 366 : 365;
  for (let m = 0; m < month - 1; m++) {
    days += MONTH_DAYS[m];
    if (m === 1 && isLeap(year)) days += 1;
  }
  days += day - 1;

  return days * 86400 + hour * 3600 + min * 60 + sec;
}
